strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,353 @@
+"""
+# pylint: disable=too-many-instance-attributes,attribute-defined-outside-init
+
+from operator import itemgetter
+from math import ceil
+import strawberryfields as sf
+from strawberryfields import ops
+from strawberryfields.parameters import par_is_symbolic
+
+
+def shift_by(l, n):
+    """Convenience function to shift a list by a number of steps.
+
+    If ``n`` is positive it shifts to the left.
+
+    Args:
+        l (list): list to be shifted
+        n (int): size of the shift
+    """
+    return l[n:] + l[:n]
+
+
+def get_modes(cmd, q):
+    """Returns the registers from q required by a command.
+
+    Args:
+        cmd (.Command): a Strawberryfields command
+        q (.RegRef): a register object
+
+    Return:
+        modes (tuple[int]): sequence of mode labels
+    """
+    reg_getter = itemgetter(*[r.ind for r in cmd.reg])
+    modes = reg_getter(q)
+
+    if not isinstance(modes, tuple):
+        modes = (modes,)
+
+    return modes
+
+
+def validate_measurements(circuit, N):
+    """Validate the TDM program measurements are correct.
+
+    Args:
+        circuit (list): a list containing commands or measurements specifying a circuit
+        N (list): list giving the number of concurrent modes per band
+
+    Return:
+        spatial_modes (int): number of spatial modes
+    """
+    spatial_modes = 0
+    measurement_reg = []
+
+    for cmd in circuit:
+        if isinstance(cmd.op, ops.Measurement):
+            modes = [r.ind for r in cmd.reg]
+            measurement_reg.extend(modes)
+            spatial_modes += len(modes)
+
+    if not spatial_modes:
+        raise ValueError("Must be at least one measurement.")
+
+    if spatial_modes is not len(N):
+        raise ValueError("Number of measurement operators must match number of spatial modes.")
+
+    return spatial_modes
+
+
+def input_check(args, copies):
+    """Checks the input arguments have consistent dimensions.
+
+    Args:
+        args (Sequence[Sequence]): sequence of sequences specifying the value of the parameters
+        copies (int): number of times the circuit should be run
+
+    """
+    # check if all lists are of equal length
+    param_lengths = [len(param) for param in args]
+    if len(set(param_lengths)) != 1:
+        raise ValueError("Gate-parameter lists must be of equal length.")
+
+    # check copies
+    if not isinstance(copies, int) or copies < 1:
+        raise TypeError("Number of copies must be a positive integer.")
+
+
+def _get_mode_order(num_of_values, N):
+    """Get the order by which the modes were measured"""
+    N = N.copy()
+
+    all_modes = []
+    for i in range(len(N)):
+        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
+        all_modes.append(ra * ceil(max(N) / len(ra)))
+
+    mode_order = [i for j in zip(*all_modes) for i in j]
+    mode_order *= ceil(num_of_values / len(mode_order))
+
+    return mode_order[:num_of_values]
+
+
+def reshape_samples(all_samples, modes, N):
```

Comment:
A note on the `reshape_samples` function. Currently it changes `all_samples`, leaving it empty, and returns a new `new_samples` that then replaces the, now empty, `all_samples` (inside `engine.py`). Thinking about it, this might not be the best solution. Either a deep copy could be made of `all_samples` in the beginning, but this could be computationally heavy if having many samples, or `new_samples` can replace `all_samples` instead of returning it, which might just make more sense than what's happening now. :thinking: 

What do you think @nquesada, @josh146?

Reply:
I'd prefer the one that minimizes computational costs given that the number of samples involved here can be massive.

Reply:
> or new_samples can replace all_samples instead of returning it

The only way I can see this working is if `reshape_samples` is either a method of the engine, _or_ it accepts and modifies an engine object directly? In either case though, I'm happy with this function for now 🙂 I agree with @nquesada - I would favour computational efficiency with some weird behaviour over more elegant behaviour with an added cost